File: newShiny/pitchSplit.R
library(dplyr)
library(tidyr)
library(tibble)
library(ggplot2)
library(cowplot)

# This function calculates the percentage of each pitch type thrown in every ball-strike count
# using either Trackman or human tagging, as specified by the 'tag' parameter
get_pitch_type_percentages <- function(pitch_data, tag) {
  
  # Filter out rows with missing count info or undefined pitch types
  pitch_data <- pitch_data %>%
    filter(!is.na(balls) & !is.na(strikes) & !is.na(.data[[tag]]) & .data[[tag]] != "Undefined")
  
  # If there's no usable data after filtering, return an empty data frame with expected columns
  if (nrow(pitch_data) == 0) {
    return(data.frame(Pitch_Type = character(0), Count = character(0), Percentage = numeric(0)))
  }
  
  # Create a "Count" column combining balls and strikes in the format "B-S"
  pitch_data$Count <- paste(pitch_data$balls, "-", pitch_data$strikes)
  
  # - Count the number of each pitch type thrown in each count
  pitch_summary <- pitch_data %>%
    group_by(Count, .data[[tag]]) %>%
    summarise(pitch_count = n(), .groups = "drop") %>%
    group_by(Count) %>%
    mutate(total_pitches_in_count = sum(pitch_count)) %>%
    ungroup() %>%
    spread(key = .data[[tag]], value = pitch_count, fill = 0) %>%
    # Round percentages to 1 decimal place
    mutate(across(-Count & -total_pitches_in_count, ~ round(. / total_pitches_in_count * 100, 1))) %>%
    arrange(Count) %>%
    dplyr::select(-any_of("total_pitches_in_count"))
  
  return(pitch_summary)
  
}

File: python_app/analysis/pitch_split.py
"""
Pitch usage analysis.

Calculates pitch type percentages for each ball-strike count.
Equivalent to pitchSplit.R.
"""

import pandas as pd


def get_pitch_type_percentages(pitch_data, tag):
    """
    Calculate the percentage of each pitch type thrown in every ball-strike count.

    Args:
        pitch_data: DataFrame with pitch data including 'balls', 'strikes', and the tag column.
        tag: Column name for pitch classification ('auto_pitch_type' or 'tagged_pitch_type').

    Returns:
        DataFrame with 'Count' column and one column per pitch type containing percentages.
    """
    if pitch_data is None or pitch_data.empty:
        return pd.DataFrame(columns=["Count"])

    df = pitch_data.copy()

    # Filter out rows with missing count info or undefined pitch types
    df = df.dropna(subset=["balls", "strikes", tag])
    df = df[df[tag] != "Undefined"]

    if df.empty:
        return pd.DataFrame(columns=["Count"])

    # Create Count column in "B - S" format
    df["Count"] = df["balls"].astype(str) + " - " + df["strikes"].astype(str)

    # Count pitches per count per pitch type
    grouped = df.groupby(["Count", tag]).size().reset_index(name="pitch_count")

    # Calculate totals per count
    totals = grouped.groupby("Count")["pitch_count"].transform("sum")

    # Calculate percentages
    grouped["pct"] = (grouped["pitch_count"] / totals * 100).round(1)

    # Pivot to wide format: one column per pitch type
    result = grouped.pivot_table(
        index="Count", columns=tag, values="pct", fill_value=0
    ).reset_index()

    # Sort by count
    result = result.sort_values("Count").reset_index(drop=True)

    # Flatten column names
    result.columns.name = None

    return result
